_P248_TempHumidity_AHT1x.ino

```diff
@@ -0,0 +1,339 @@
+  //  return false;
+  //}
+
+  //if (Wire.endTransmission(true) != 0) return AHT10_ERROR;             //safety check, make sure transmission complete
+
+  //return 0x00;
+
+}
+
+
+
+//==============================================
+// PLUGIN
+// =============================================
+
+boolean Plugin_248(byte function, struct EventStruct *event, String& string)
+{
+  boolean success = false;
+
+  switch (function)
+  {
+    case PLUGIN_DEVICE_ADD:
+    {
+      Device[++deviceCount].Number = PLUGIN_ID_248;
+      Device[deviceCount].Type = DEVICE_TYPE_I2C;
+      Device[deviceCount].VType = Sensor_VType::SENSOR_TYPE_TEMP_HUM;
+      Device[deviceCount].Ports = 0;
+      Device[deviceCount].PullUpOption = false;
+      Device[deviceCount].InverseLogicOption = false;
+      Device[deviceCount].FormulaOption = true;
+      Device[deviceCount].ValueCount = 2;
+      Device[deviceCount].SendDataOption = true;
+      Device[deviceCount].TimerOption = true;
+      Device[deviceCount].GlobalSyncOption = true;
+      break;
+    }
+
+    case PLUGIN_GET_DEVICENAME:
+    {
+      string = F(PLUGIN_NAME_248);
+      break;
+    }
+
+    case PLUGIN_GET_DEVICEVALUENAMES:
+    {
+      strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[0], PSTR(PLUGIN_VALUENAME1_248));
+      strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[1], PSTR(PLUGIN_VALUENAME2_248));
+      //strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[2], PSTR(PLUGIN_VALUENAME3_248));
+      //strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[3], PSTR(PLUGIN_VALUENAME4_248));
+    //  strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[4], PSTR(PLUGIN_VALUENAME5_248));
+      break;
+    }
+
+    case PLUGIN_WEBFORM_SHOW_I2C_PARAMS:
+    {
+      byte choice = PCONFIG(0);
+
+      int optionValues[2] = { 0x38, 0x39 };
+      addFormSelectorI2C(F("i2c_addr"), 2, optionValues, choice);
+      addFormNote(F("AO Low=0x38, High=0x39"));
+      break;
+    }
+
+    case PLUGIN_WEBFORM_LOAD:
+    {
+      success = true;
+      break;
+    }
+
+    case PLUGIN_WEBFORM_SAVE:
+    {
+      PCONFIG(0) = getFormItemInt(F("i2c_addr"));
+
+      success = true;
+      break;
+    }
+
+    case PLUGIN_INIT:
+    {
+      String log;
+      uint8_t addr = PCONFIG(0);
+
+      if (!AHT10_init(addr))
+        log = F("AHT10: init");
+
+      //if (!AHT10_setNormalMode(addr))         //one measurement+sleep mode
+      //  log = F("AHT10: setNormalMode");
+      //if (!AHT10_enableFactoryCalCoeff(addr)) //load factory calibration coeff
+       // log = F("AHT10: enableFactoryCoeff");
+
+      /*check calibration enable */
+      //if (AHT10_getCalibrationBit(false) != 0x01) 
+       // log = F("AHT10: getCalibrationBit");
+
+      if (log)
+        addLog(LOG_LEVEL_ERROR,log);
+
+      success = true;
+      break;
+    }
+
+    case PLUGIN_READ:
+    {
+      uint8_t addr = PCONFIG(0);
+      AHT1XStartMeasurement(addr);
+      delay(AHT10_MEASURMENT_DELAY);
```

Comment:
Nope!
Such a delay on each read call is a big no-no.
Rather perform the start and set some new to be created plugin state (for this plugin) to "started".
Then re-schedule this plugin to read again (see for example BME280's implementation) and check the state to see a measurement should be ready.

Reply:
Hmm. The BME is schedule for a read every second. AHT10 can be heated if read too often. 
Can I do a schedule from read back to read ?
Seems strange to need to do this.

Reply:
can I use the PLUGIN_TIMER_IN from the read ? Any way I can pass some parameters to the timer task ?

Reply:
it looks like we try to implement a state machine by trying to avoid the delay() method at all costs. 
May I know the reason behind?

Reply:
A call to `delay` is blocking.
Since we try to use as many tasks/controllers on this microprocessor which all run on the same loop, we must avoid blocking code as that may lead to lots of issues.
For example, lost WiFi connections, missing requests (timeouts) and scheduled actions not taking place at the intended interval.
Also we won't serve HTTP/DNS calls and don't serve other incoming packets, which may lead to overflowing buffers and queues.

I've spent way over several 100's of hours (perhaps > 1000h) trying to get rid of these issues, to make ESPEasy as responsive and stable as possible, so I will try to hunt down every delay over a few msec in length in the code.

Please have a look at the timing stats page.
Every blocking code > 100 msec is highlighted and marked in bold.
It would be best to have all values running regularly to be as small as possible, preferrably < 10 msec.
Serving a web page may take a bit more, but nowadays most web pages we serve have a load time of 150 - 200 msec on ESP8266 and way less on ESP32 with ethernet.

Reply:
> can I use the PLUGIN_TIMER_IN from the read ? Any way I can pass some parameters to the timer task ?

You can't give this information along with the event, but you can store task specific runtime information in the taskdata structure I mentioned before (see P107 for a very clean example, or any plugin calling `getPluginTaskData`)
The state of a task can be stored in the plugin task data array.

Reply:
> Hmm. The BME is schedule for a read every second. AHT10 can be heated if read too often.
> Can I do a schedule from read back to read ?
> Seems strange to need to do this.

`PLUGIN_READ` is called only every "interval" number of seconds.
So what you can do, is start your measurement in the first call of `PLUGIN_READ` and then set the state accordingly.
You also reschedule a new call for `PLUGIN_READ` just as is done in P028_BME280:

```c++
Scheduler.schedule_task_device_timer(event->TaskIndex, millis() + 10); // Add your own expected offset, 10 msec in this example
```

So in that second call of `PLUGIN_READ` you read the measurement and set the state according.
The scheduler automatically will schedule a call to `PLUGIN_READ` after set "interval" nr of seconds.
By re-scheduling you cause some drift in delivering data, so you could try to compensate for that (as I've done in the Dallas plugin), but I don't think it really matters.

Reply:
Thank you for the explanation.
Currently working on removing that delay and use the same approach as the BME280 implementation.